### PR TITLE
TDK-7908: L2 Tests for dsDisplay and dsCompositeIn

### PR DESCRIPTION
**Pull request by saniya-09**
Reason for change: Generate L2 tests for dsDisplay and dsCompositeIn
Risks: Low
Test Procedure: Execute the hal test binary

### CODE REVIEW

File: docs/pages/l2_dsCompositeIn_test_specification.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+**User Interaction :**
+If user chose to run the test in interactive mode, then the test case has to be selected via console.
+
+#### Test Procedure :
+
+| Variation / Steps | Description | Test Data | Expected Result | Notes|
+| -- | --------- | ---------- | -------------- | ----- |
+| 01 | Initialize the COMPOSITE Input ports using dsCompositeInInit | None | dsERR_NONE | Should be successful |
+| 02 | Get the number of COMPOSITE Input ports using dsCompositeInGetNumberOfInputs | numInputs = valid buffer | dsERR_NONE | Should be successful |
+| 03 | Verify the returned number of COMPOSITE Input ports | numInputs = valid buffer | numInputs >= 0 | Should be successful |
```
Comment (srinivasgtl) [inline docs/pages/l2_dsCompositeIn_test_specification.md:26]: Verification should be done with the value read from the config file.  The above statement doesn't say with what should the numInputs should be validated.

File: docs/pages/l2_dsCompositeIn_test_specification.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+**User Interaction :**
+If user chose to run the test in interactive mode, then the test case has to be selected via console.
+
+#### Test Procedure :
+
+| Variation / Steps | Description | Test Data | Expected Result | Notes|
+| -- | --------- | ---------- | -------------- | ----- |
+| 01 | Initialize the Composite Input using dsCompositeInInit() | None | dsERR_NONE | Should be successful |
+| 02 | Get the status of the Composite Input using dsCompositeInGetStatus() | status = valid buffer | dsERR_NONE | Should be successful |
+| 03 | Verify if the Composite Input is disabled | status.isPresented = false | dsERR_NONE | Should be successful |
```
Comment (srinivasgtl) [inline docs/pages/l2_dsCompositeIn_test_specification.md:68]: Other elements of the buffer also need to be validated.  status.isPortConnected, isActivePort etc.

File: docs/pages/l2_dsCompositeIn_test_specification.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+**User Interaction :**
+If user chose to run the test in interactive mode, then the test case has to be selected via console.
+
+#### Test Procedure :
+
+| Variation / Steps | Description | Test Data | Expected Result | Notes|
+| -- | --------- | ---------- | -------------- | ----- |
+| 01 | Initialize the Composite Input using dsCompositeInInit() | None | dsERR_NONE | Should be successful |
+| 02 | Select the Composite Input port using dsCompositeInSelectPort() | port = dsCOMPOSITE_IN_PORT_0 | dsERR_NONE | Should be successful |
+| 03 | Verify the selected port is active (implementation dependent) | activePort = port | activePort should be equal to port | Should be successful |
```
Comment (srinivasgtl) [inline docs/pages/l2_dsCompositeIn_test_specification.md:109]: activePort should be false and not "port"

**Review by srinivasgtl — COMMENTED** [review]
Heading should be added.
Table of contents, Acronyms etc should be added.

**Review by srinivasgtl — COMMENTED** [review]
Heading should be added.
Contents and Acronyms etc should be added.

File: src/test_l2_dsCompositeIn.c [hunk 4]
```diff
@@ -1,119 +1,307 @@
...
+    {
+        UT_LOG("dsCompositeInGetNumberOfInputs failed with status: %d", status);
+        // Call the post-requisite API dsCompositeInTerm to terminate the COMPOSITE Input module.
+        dsCompositeInTerm();
+        return;
+    }
+
+    // Step 4: If the return status is dsERR_NONE, compare the value of the uint8_t variable with the expected number of COMPOSITE input ports from the configuration file.
+    UT_ASSERT_EQUAL(numInputs, expectedNumInputs);
+    if (numInputs != expectedNumInputs)
```
Comment (bhanucbp) [inline src/test_l2_dsCompositeIn.c:83]: this if condition is not required as the previous line assert will check the condition

File: src/test_l2_dsCompositeIn.c [hunk 5]
```diff
@@ -1,119 +1,307 @@
...
+    UT_ASSERT_EQUAL(ret, dsERR_NONE);
+    if (ret != dsERR_NONE)
+    {
+        UT_LOG("dsCompositeInInit failed with error: %d", ret);
+        return;
+    }
+
+    // Select a random port from the available ports
+    srand(time(NULL));
+    port = rand() % dsCOMPOSITE_IN_PORT_MAX;
```
Comment (bhanucbp) [inline src/test_l2_dsCompositeIn.c:191]: dsCompositeInGetNumberOfInputs needs to be called to get the supported composite in ports and use that value to loop all the ports. Spec is missing this information. Spec needs to be updated

File: src/test_l2_dsCompositeIn.c [hunk 6]
```diff
@@ -1,119 +1,307 @@
...
+
+    status = dsCompositeInInit();
+    UT_ASSERT_EQUAL(status, dsERR_NONE);
+    if (status != dsERR_NONE)
+    {
+        UT_LOG("dsCompositeInInit failed with status: %d", status);
+        return;
+    }
+
+    status = dsCompositeInScaleVideo(x, y, width, height);
```
Comment (bhanucbp) [inline src/test_l2_dsCompositeIn.c:263]: dsCompositeInSelectPort is pre requisite  for dsCompositeInScaleVideo. But this API not called

File: src/test_l2_dsDisplay.c [hunk 7]
```diff
@@ -1,121 +1,172 @@
...
+
+    // Step 1
+    ret = dsDisplayInit();
+    UT_LOG("Invoked dsDisplayInit(). Return status: %d", ret);
+    UT_ASSERT_EQUAL(ret, dsERR_NONE);
+
+    if (ret == dsERR_NONE)
+    {
+        // Step 2
+        ret = dsGetDisplay(dsVIDEOPORT_TYPE_HDMI, 0, &handle);
```
Comment (bhanucbp) [inline src/test_l2_dsDisplay.c:68]: for sink devices display should be dsVIDEOPORT_TYPE_INTERNAL. dsVIDEOPORT_TYPE_HDMI is not supported. Spec needs to be updated with details

File: src/test_l2_dsDisplay.c [hunk 8]
```diff
@@ -1,121 +1,172 @@
...
+* **Test Group ID:** 02@n
+* **Test Case ID:** 002@n
 *
 * **Test Procedure:**
-* TODO: Refer to L2 Specification documentation : [l2_module_test_specification_template.md](../docs/pages/l2_module_test_specification_template.md)
+* Refer to UT specification documentation [l2_dsDisplay_test_specification.md](l2_dsDisplay_test_specification.md)
 */
-void test_l2_dsDisplay (void)
+
+void test_l2_dsDisplay_TestDefaultAspectRatio(void)
```
Comment (bhanucbp) [inline src/test_l2_dsDisplay.c:115]: as per the spec test_l2_dsDisplay_ValidateEDID is only for sink devices and test_l2_dsDisplay_TestDefaultAspectRatio is only for source devices. But there is no way we can differentiate. Its better to have some postfix (_sink, _source) to differentiate

File: docs/pages/l2_dsCompositeIn_test_specification.md (hunk 2 above)
Comment (saniya-09) [reply docs/pages/l2_dsCompositeIn_test_specification.md]: Done

File: docs/pages/l2_dsCompositeIn_test_specification.md (hunk 3 above)
Comment (saniya-09) [reply docs/pages/l2_dsCompositeIn_test_specification.md]: Done

File: src/test_l2_dsCompositeIn.c (hunk 5 above)
Comment (saniya-09) [reply src/test_l2_dsCompositeIn.c]: Done

File: src/test_l2_dsCompositeIn.c (hunk 6 above)
Comment (saniya-09) [reply src/test_l2_dsCompositeIn.c]: This testcase is expected to call dsCompositeInScaleVideo without calling dsCompositeInSelectPort api.

File: docs/pages/l2_dsCompositeIn_test_specification.md (hunk 1 above)
Comment (smrithisurendran) [reply docs/pages/l2_dsCompositeIn_test_specification.md]: to be addressed with profile support upgrade

File: src/test_l2_dsCompositeIn.c (hunk 4 above)
Comment (smrithisurendran) [reply src/test_l2_dsCompositeIn.c:96]: this is to ensure that test exits after cleanup on assertion failure, as agreed on previous discussions

File: docs/pages/l2_dsCompositeIn_test_specification.md (hunk 1 above)
Comment (saniya-09) [reply docs/pages/l2_dsCompositeIn_test_specification.md]: Done

File: docs/pages/l2_dsDisplay_test_specification.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+**User Interaction :**
+If user chose to run the test in interactive mode, then the test case has to be selected via console.
+
+#### Test Procedure :
+
+| Variation / Steps | Description | Test Data | Expected Result | Notes|
+| -- | --------- | ---------- | -------------- | ----- |
+| 01 | Initialize DS Display using dsDisplayInit | None | dsERR_NONE | Should be successful |
+| 02 | Get Display Handle using dsGetDisplay | dsVIDEOPORT_TYPE_HDMI, 0, &handle | dsERR_NONE | Should be successful |
+| 03 | Validate EDID using dsGetEDID | handle, &edid | dsERR_NONE | Should be successful |
```
[the remark this answers is not in the document]
Comment (saniya-09) [reply docs/pages/l2_dsDisplay_test_specification.md]: Product code from edid data is addressed as part of the profile . edidbytes is handled by assuming the existing functions will support and Dev will handle this manually once the array support in KVP is ready

File: src/test_l2_dsDisplay.c (hunk 7 above)
Comment (saniya-09) [reply src/test_l2_dsDisplay.c]: Done

File: src/test_l2_dsDisplay.c (hunk 8 above)
Comment (saniya-09) [reply src/test_l2_dsDisplay.c]: Done

File: src/test_l2_dsCompositeIn.c [hunk 10]
```diff
@@ -1,119 +1,214 @@
...
+    UT_LOG_INFO("In %s [%02d%03d]\n", __FUNCTION__, gTestGroup, gTestID);
+
+    dsError_t ret;
+    dsCompositeInStatus_t status;
+
+    ret = dsCompositeInInit();
+    UT_LOG_DEBUG("Invoking dsCompositeInInit()");
+    UT_ASSERT_EQUAL_FATAL(ret, dsERR_NONE);
+
+    for (dsCompositeInPort_t port = dsCOMPOSITE_IN_PORT_0; port < dsCOMPOSITE_IN_PORT_MAX; port++)
```
Comment (srinivasgtl) [inline src/test_l2_dsCompositeIn.c:161]: Should read the supported ports from the profile rather than considering the max supported?

**Comment by srinivasgtl** [comment]
These changes are now considered on other PR
Display: https://github.com/rdkcentral/rdk-halif-test-device_settings/pull/13
Composite:https://github.com/rdkcentral/rdk-halif-test-device_settings/pull/16

**Pull request closed without merging**